staging/src/k8s.io/client-go/tools/cache/reflector_data_consistency_detector.go

```diff
@@ -32,42 +33,46 @@ import (
 	"k8s.io/klog/v2"
 )
 
-var dataConsistencyDetectionEnabled = false
+var dataConsistencyDetectionForWatchListEnabled = false
 
 func init() {
-	dataConsistencyDetectionEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_WATCHLIST_INCONSISTENCY_DETECTOR"))
+	dataConsistencyDetectionForWatchListEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_WATCHLIST_INCONSISTENCY_DETECTOR"))
 }
 
-// checkWatchListConsistencyIfRequested performs a data consistency check only when
+type retrieveCollectedItemsFunc[U any] func() []U
+
+type listFunc[T runtime.Object] func(ctx context.Context, options metav1.ListOptions) (T, error)
+
+// checkWatchListDataConsistencyIfRequested performs a data consistency check only when
 // the KUBE_WATCHLIST_INCONSISTENCY_DETECTOR environment variable was set during a binary startup.
 //
 // The consistency check is meant to be enforced only in the CI, not in production.
 // The check ensures that data retrieved by the watch-list api call
-// is exactly the same as data received by the standard list api call.
+// is exactly the same as data received by the standard list api call against etcd.
 //
 // Note that this function will panic when data inconsistency is detected.
 // This is intentional because we want to catch it in the CI.
-func checkWatchListConsistencyIfRequested(stopCh <-chan struct{}, identity string, lastSyncedResourceVersion string, listerWatcher Lister, store Store) {
-	if !dataConsistencyDetectionEnabled {
+func checkWatchListDataConsistencyIfRequested[T runtime.Object, U any](ctx context.Context, identity string, lastSyncedResourceVersion string, listFn listFunc[T], retrieveCollectedItemsFn retrieveCollectedItemsFunc[U]) {
+	if !dataConsistencyDetectionForWatchListEnabled {
 		return
 	}
-	checkWatchListConsistency(stopCh, identity, lastSyncedResourceVersion, listerWatcher, store)
+	// for informers we pass an empty ListOptions because
+	// listFn might be wrapped for filtering during informer construction.
```

Comment:
for informers we can, because informers wrap `options` during construction time and we don't have access to the original params, e.g.
```
&cache.ListWatch{
	ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
		if tweakListOptions != nil {
			tweakListOptions(&options)
		}
		return client.AppsV1().Deployments(namespace).List(context.TODO(), options)
	},
...
},
```

we could do the same (wrap original options) for standard list (and potentially for other types of request) as well, or we could simply pass original `ListOptions` to the checker.

Reply:
For informers ? We don't have access to the original `ListOptions` because they are wrapped into `ListWatcher` - https://github.com/kubernetes/kubernetes/blob/9161302e7fd3a5fb055b2f2572c6e1228240bb51/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L246

We could change that but that would be a quite significant refactoring.

Reply:
We can't use the options you linked as they are invalid for list requests. The options we are interested in (those that set filters) are defined during informers construction and are wrapped into a ListWatcher. Does it make sense ?

Reply:
So you're saying that reflector listerwatcher is e.g. tweaking selectors - that's right.
I think we can make it cleaner, but maybe let's not block on it for now.